cdap-security/src/main/java/co/cask/cdap/security/tools/GeneratedCertKeyStores.java

```diff
@@ -0,0 +1,156 @@
+  private static final int VALIDITY = 999;
+
+  /* private constructor */
+  private GeneratedCertKeyStores() {}
+
+  /**
+   * Create a Java key store with a stored self-signed certificate.
+   * @return Java keystore which has a self signed X.509 certificate
+   */
+  public static KeyStore getSSLKeyStore(SConfiguration sConf) {
+    KeyStore keyStore;
+    String password = sConf.get(Constants.Security.Ssl.SSL_KEYSTORE_PASSWORD);
+    try {
+      String keyPairAlgorithm = sConf.get(Constants.Security.Ssl.KEY_PAIR_ALGORITHM,
+                                          KEY_PAIR_ALGORITHM);
+      KeyPairGenerator keyGen = KeyPairGenerator.getInstance(keyPairAlgorithm);
+      String randomAlgorithm = sConf.get(Constants.Security.Ssl.SECURE_RANDOM_ALGORITHM,
+                                         SECURE_RANDOM_ALGORITHM);
+      String randomProvider = sConf.get(Constants.Security.Ssl.SECURE_RANDOM_PROVIDER,
+                                        SECURE_RANDOM_PROVIDER);
+      SecureRandom random = SecureRandom.getInstance(randomAlgorithm, randomProvider);
+      keyGen.initialize(KEY_SIZE, random);
+      // generate a key pair
+      KeyPair pair = keyGen.generateKeyPair();
+      int validity = sConf.getInt(Constants.Security.Ssl.CERT_VALIDITY, VALIDITY);
+      String distinguishedName = sConf.get(Constants.Security.Ssl.CERT_DISTINGUISHED_NAME, DISTINGUISHED_NAME);
+      String signatureAlgo = sConf.get(Constants.Security.Ssl.SIGNATURE_ALGORITHM, SIGNATURE_ALGORITHM);
+
+      X509Certificate cert = getCertificate(distinguishedName, pair, validity, signatureAlgo);
+
+      keyStore = KeyStore.getInstance(sConf.get(Constants.Security.Ssl.SSL_KEYSTORE_TYPE, SSL_KEYSTORE_TYPE));
+      keyStore.load(null, password.toCharArray());
+      keyStore.setKeyEntry(CERT_ALIAS, pair.getPrivate(), password.toCharArray(),
+                           new java.security.cert.Certificate[]{cert});
+    } catch (Throwable e) {
+      throw new RuntimeException("SSL is enabled but a key store file could not be created. A keystore is required " +
+                                   "for SSL to be used.", e);
+    }
+    return keyStore;
+  }
+
+  /**
+   * Generate an X.509 certificate
+   *
+   * @param dn Distinguished name for the owner of the certificate, it will also be the signer of the certificate.
+   * @param pair Key pair used for signing the certificate.
+   * @param days Validity of the certificate.
+   * @param algorithm Name of the signature algorithm used.
+   * @return A X.509 certificate
+   */
+  private static X509Certificate getCertificate(String dn, KeyPair pair, int days, String algorithm) {
+    // Calculate the validity interval of the certificate
+    Date from = new Date();
+    Date to = DateUtils.addDays(from, days);
+    CertificateValidity interval = new CertificateValidity(from, to);
+    // Generate a random number to use as the serial number for the certificate
+    BigInteger sn = new BigInteger(64, new SecureRandom());
```

Comment:
Do you know what happens if two certificates have the same serial number? What is the purpose of the serial number?